lib/services/parser.go

```diff
@@ -731,136 +732,69 @@ func newParserForIdentifierSubcondition(ctx RuleContext, identifier string) (pre
 //   - shorthand `labels` refers to resource `resource.metadata.labels + resource.spec.dynamic_labels`
 //     eg: `labels.env == "prod"`
 //
 // All other fields can be referenced by starting expression with identifier `resource`
 // followed by the names of the json fields ie: `resource.spec.public_addr`.
-func NewResourceParser(resource types.ResourceWithLabels) (BoolPredicateParser, error) {
-	predEquals := func(a interface{}, b interface{}) predicate.BoolPredicate {
-		switch aval := a.(type) {
-		case label:
-			bval, ok := b.(string)
-			return func() bool {
-				return ok && aval.value == bval
-			}
-		default:
-			return predicate.Equals(a, b)
-		}
-	}
-	predPrefix := func(a interface{}, prefix string) predicate.BoolPredicate {
-		switch aval := a.(type) {
-		case label:
-			return func() bool {
-				return strings.HasPrefix(aval.value, prefix)
-			}
-		case string:
-			return func() bool {
-				return strings.HasPrefix(aval, prefix)
-			}
-		default:
-			return func() bool {
-				return false
-			}
-		}
-	}
-
-	p, err := predicate.NewParser(predicate.Def{
-		Operators: predicate.Operators{
-			AND: predicate.And,
-			OR:  predicate.Or,
-			NOT: predicate.Not,
-			EQ:  predEquals,
-			NEQ: func(a interface{}, b interface{}) predicate.BoolPredicate {
-				return predicate.Not(predEquals(a, b))
-			},
-		},
-		Functions: map[string]interface{}{
-			"hasPrefix": predPrefix,
-			"equals":    predEquals,
-			// search allows fuzzy matching against select field values.
-			"search": func(searchVals ...string) predicate.BoolPredicate {
-				return func() bool {
-					return resource.MatchSearch(searchVals)
-				}
-			},
-			// exists checks for an existence of a label by checking
-			// if a key exists. Label value are unchecked.
-			"exists": func(l label) predicate.BoolPredicate {
-				return func() bool {
-					return len(l.key) > 0
-				}
-			},
-		},
-		GetIdentifier: func(fields []string) (interface{}, error) {
-			switch fields[0] {
-			case ResourceLabelsIdentifier:
-				switch {
-				// Field length of 1 means the user is using
-				// an index expression ie: labels["env"], which the
-				// parser will expect a map for lookup in `GetProperty`.
-				case len(fields) == 1:
-					return resource, nil
-				case len(fields) > 2:
-					return nil, trace.BadParameter("only two fields are supported with identifier %q, got %d: %v", ResourceLabelsIdentifier, len(fields), fields)
-				default:
-					key := fields[1]
-					val, ok := resource.GetLabel(key)
-					if ok {
-						return label{key: key, value: val}, nil
-					}
-					return label{}, nil
-				}
-
-			case ResourceNameIdentifier:
-				if len(fields) > 1 {
-					return nil, trace.BadParameter("only one field are supported with identifier %q, got %d: %v", ResourceNameIdentifier, len(fields), fields)
-				}
-
+func NewResourceParser(expression string) (typical.Expression[types.ResourceWithLabels, bool], error) {
```

Comment:
I think `NewResourceParser` is a misnomer now that this returns an already parsed expression rather than a parser